contracts/StablecoinConverter.sol

```diff
@@ -167,4 +174,49 @@ contract StablecoinConverter is EpochTokenLocker {
         }
         revert("Price not provided for token");
     }
+
+    function documentTrades(
+        uint batchIndex,
+        address[] memory owner,  //tradeData is submitted as arrays
+        uint16[] memory orderId,
+        uint128[] memory volume,
+        uint128[] memory prices,
+        uint16[] memory tokenIdForPrice
+    ) internal {
+        previousSolutionBatchId = batchIndex;
+        uint len = owner.length;
+        for (uint i = 0; i < len; i++) {
+            previousSolutionTrades.push(TradeData({
+                owner: owner[i],
+                orderId: orderId[i],
+                volume: volume[i]
+            }));
+        }
+        len = prices.length;
+        for (uint i = 0; i < len; i++) {
+            previousSolutionPrices.push(prices[i]);
+            previousSolutionTokenIdForPrice.push(tokenIdForPrice[i]);
+        }
+    }
+
+    function undoPreviousSolution(uint batchIndex) internal {
+        if (previousSolutionBatchId == batchIndex) {
+            for (uint i = 0; i < previousSolutionTrades.length; i++) {
+                Order memory order = orders[previousSolutionTrades[i].owner][previousSolutionTrades[i].orderId];
+                uint sellVolume = previousSolutionTrades[i].volume;
+                uint buyVolume = previousSolutionTrades[i].volume
+                    .mul(previousSolutionPrices[findPriceIndex(order.buyToken, previousSolutionTokenIdForPrice)]) /
+                    previousSolutionPrices[findPriceIndex(order.sellToken, previousSolutionTokenIdForPrice)];
+
+                order.buyAmount = uint128(order.sellAmount.add(sellVolume).mul(order.buyAmount) / order.sellAmount);
+                order.sellAmount = uint128(order.sellAmount.add(sellVolume));
+                orders[previousSolutionTrades[i].owner][previousSolutionTrades[i].orderId] = order;
+                subtractBalance(previousSolutionTrades[i].owner, tokenIdToAddressMap(order.buyToken), buyVolume);
```

Comment:
I think you might have to do the subtractions after the additions again since otherwise you might become temporarily negative. A unit test for this would be nice.

Reply:
I love unit tests! ❤️